File: lib/sign-outbound.js
'use strict';

const crypto = require('crypto');
const fs = require('fs');

let cachedKey;
let cachedKeyPath;

function ed25519Pkcs8(parsedKey) {
  const seed = parsedKey.part.k.data.slice(0, 32);
  if (seed.length !== 32) {
    throw new Error(`expected a 32-byte ed25519 seed, got ${seed.length}`);
  }
  const header = Buffer.from('302e020100300506032b657004220420', 'hex');
  return {
    key: Buffer.concat([header, seed]),
    format: 'der',
    type: 'pkcs8',
  };
}

function loadSigningKey() {
  const keyPath = (process.env.CS_SIGNING_KEY || '').trim();
  if (!keyPath) return null;

  if (cachedKey !== undefined && cachedKeyPath === keyPath) return cachedKey;
  cachedKeyPath = keyPath;

  try {
    const sshpk = require('sshpk');
    const parsed = sshpk.parsePrivateKey(fs.readFileSync(keyPath), 'auto');
    if (parsed.type !== 'ed25519') {
      throw new Error(`key is ${parsed.type}; only ed25519 is accepted`);
    }
    cachedKey = crypto.createPrivateKey(ed25519Pkcs8(parsed));
  } catch (err) {
    cachedKey = undefined;
    throw new Error(
      `[sign-outbound] CS_SIGNING_KEY=${keyPath} unusable: ${err.message}`
    );
  }
  return cachedKey;
}

function assertUsable() {
  loadSigningKey();
}

function signHeaders(agentId, method, path, body) {
  const key = loadSigningKey();
  if (!key) return {};

  const id = agentId || (process.env.CS_AGENT_ID || '').trim() || 'cs-core';
  const timestamp = new Date().toISOString().split('.')[0] + 'Z';
  const nonce = crypto.randomBytes(16).toString('hex');
  const cleanPath = String(path || '/').split('?')[0];
  const bodyBytes = Buffer.isBuffer(body)
    ? body
    : Buffer.from(body || '', 'utf8');
  const bodySha256 = crypto.createHash('sha256').update(bodyBytes).digest('hex');

  const canonical = [
    String(method || '').toUpperCase(),
    cleanPath,
    timestamp,
    nonce,
    bodySha256,
  ].join('\n');

  const signature = crypto
    .sign(null, Buffer.from(canonical, 'utf8'), key)
    .toString('base64');

  return {
    'X-Agent-Id': id,
    'X-Timestamp': timestamp,
    'X-Nonce': nonce,
    'X-Signature': signature,
  };
}

module.exports = { signHeaders, assertUsable };

File: lib/sign_outbound.py
"""Signs outbound service-to-service calls. Protocol: docs/SIGNING-PROTOCOL.md."""

import base64
import hashlib
import os
import secrets
from datetime import datetime, timezone

_key = None
_key_path = None


def _load_key():
    """Load and cache the signing key. Raises if configured but unusable."""
    global _key, _key_path

    path = (os.environ.get("CS_SIGNING_KEY") or "").strip()
    if not path:
        return None
    if _key is not None and _key_path == path:
        return _key or None
    _key_path = path

    try:
        from cryptography.hazmat.primitives.serialization import (
            load_ssh_private_key,
        )
        from cryptography.hazmat.primitives.asymmetric.ed25519 import (
            Ed25519PrivateKey,
        )

        with open(path, "rb") as fh:
            key = load_ssh_private_key(fh.read(), password=None)
        if not isinstance(key, Ed25519PrivateKey):
            raise TypeError(
                f"key is {type(key).__name__}; only ed25519 is accepted"
            )
        _key = key
    except Exception as err:
        _key = None
        raise RuntimeError(
            f"[sign-outbound] CS_SIGNING_KEY={path} unusable: {err}"
        ) from err
    return _key


def assert_usable():
    """Raises at startup if a key is configured but unloadable, instead of per-request 401s."""
    _load_key()


def agent_id() -> str:
    return (os.environ.get("CS_AGENT_ID") or "").strip() or "cs-core"


def is_configured() -> bool:
    return bool((os.environ.get("CS_SIGNING_KEY") or "").strip())


def sign_headers(method: str, path: str, body=b"") -> dict:
    """Builds the four signing headers; `path` is as the callee sees it, query stripped.

    Returns {} when signing is not configured.
    """
    key = _load_key()
    if key is None:
        return {}

    if isinstance(body, str):
        body = body.encode("utf-8")
    body = body or b""

    timestamp = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
    nonce = secrets.token_hex(16)
    clean_path = (path or "/").split("?")[0]
    body_sha256 = hashlib.sha256(body).hexdigest()

    canonical = "\n".join([
        (method or "").upper(),
        clean_path,
        timestamp,
        nonce,
        body_sha256,
    ])
    signature = base64.b64encode(
        key.sign(canonical.encode("utf-8"))
    ).decode("ascii")

    return {
        "X-Agent-Id": agent_id(),
        "X-Timestamp": timestamp,
        "X-Nonce": nonce,
        "X-Signature": signature,
    }
